# Using Pandas

In [4]:
import pandas as pd
import numpy as np
pd.set_option('display.max_rows', 200)
## to make it possible to display multiple output inside one cell 
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

<b>load the data from the vehicles.csv file into pandas data frame

In [5]:
## Your Code here

vehicles_df = pd.read_csv("data/vehicles.csv")

First exploration of the dataset:

- How many observations does it have?
- Look at all the columns: do you understand what they mean?
- Look at the raw data: do you see anything weird?
- Look at the data types: are they the expected ones for the information the column contains?

In [6]:
## Your Code here

vehicles_df.info()
vehicles_df.count()
vehicles_df.describe()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35952 entries, 0 to 35951
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Make                     35952 non-null  object 
 1   Model                    35952 non-null  object 
 2   Year                     35952 non-null  int64  
 3   Engine Displacement      35952 non-null  float64
 4   Cylinders                35952 non-null  float64
 5   Transmission             35952 non-null  object 
 6   Drivetrain               35952 non-null  object 
 7   Vehicle Class            35952 non-null  object 
 8   Fuel Type                35952 non-null  object 
 9   Fuel Barrels/Year        35952 non-null  float64
 10  City MPG                 35952 non-null  int64  
 11  Highway MPG              35952 non-null  int64  
 12  Combined MPG             35952 non-null  int64  
 13  CO2 Emission Grams/Mile  35952 non-null  float64
 14  Fuel Cost/Year        

Make                       35952
Model                      35952
Year                       35952
Engine Displacement        35952
Cylinders                  35952
Transmission               35952
Drivetrain                 35952
Vehicle Class              35952
Fuel Type                  35952
Fuel Barrels/Year          35952
City MPG                   35952
Highway MPG                35952
Combined MPG               35952
CO2 Emission Grams/Mile    35952
Fuel Cost/Year             35952
dtype: int64

Year  Engine Displacement     Cylinders  Fuel Barrels/Year  \
count  35952.00000         35952.000000  35952.000000       35952.000000   
mean    2000.71640             3.338493      5.765076          17.609056   
std       10.08529             1.359395      1.755268           4.467283   
min     1984.00000             0.600000      2.000000           0.060000   
25%     1991.00000             2.200000      4.000000          14.699423   
50%     2001.00000             3.000000      6.000000          17.347895   
75%     2010.00000             4.300000      6.000000          20.600625   
max     2017.00000             8.400000     16.000000          47.087143   

           City MPG   Highway MPG  Combined MPG  CO2 Emission Grams/Mile  \
count  35952.000000  35952.000000  35952.000000             35952.000000   
mean      17.646139     23.880646     19.929322               475.316339   
std        4.769349      5.890876      5.112409               119.060773   
min        6.000000      9.000000      7.000000                37.000000   
25%       15.000000     20.000000     16.000000               395.000000   
50%       17.000000     24.000000     19.000000               467.736842   
75%       20.000000     27.000000     23.000000               555.437500   
max       58.000000     61.000000     56.000000              1269.571429   

       Fuel Cost/Year  
count    35952.000000  
mean      1892.598465  
std        506.958627  
min        600.000000  
25%       1500.000000  
50%       1850.000000  
75%       2200.000000  
max       5800.000000

###How many observations does it have?

The dataset has 35952 entries.

###Look at all the columns: do you understand what they mean?

We don´t now what year was used for Fuel Cost/Year 

###Look at the raw data: do you see anything weird?

The Minimum of CO2 Emission Grams/Mile and Fuel Barrels/Year looks too low

###Look at the data types: are they the expected ones for the information the column contains?
I would say so



### Cleaning and wrangling data

- Some car brand names refer to the same brand. Replace all brand names that contain the word "Dutton" for simply "Dutton". If you find similar examples, clean their names too. Use `loc` with boolean indexing.

- Convert CO2 Emissions from Grams/Mile to Grams/Km

- Create a binary column that solely indicates if the transmission of a car is automatic or manual. Use `pandas.Series.str.startswith` and .

- convert MPG columns to km_per_liter

In [7]:
## Some car brand names refer to the same brand. Replace all brand names that contain the word "Dutton" for simply "Dutton". If you find similar examples, clean their names too. Use loc with boolean indexing.

vehicles_df.Make.value_counts()

dutton_names = vehicles_df[vehicles_df["Make"].str.contains("Dutton")==True]
vehicles_df["Make"].replace(["E. P. Dutton, Inc.","Superior Coaches Div E.p. Dutton","S and S Coach Company  E.p. Dutton"],"Dutton", inplace=True)

PAS_names = vehicles_df[vehicles_df["Make"].str.contains("PAS")==True]
vehicles_df["Make"].replace(["PAS Inc - GMC","PAS, Inc"],"PAS Inc", inplace=True)

BMW_names = vehicles_df[vehicles_df["Make"].str.contains("BMW")==True]
vehicles_df["Make"].replace(["BMW Alpina"],"BMW", inplace=True)

Grumman_names = vehicles_df[vehicles_df["Make"].str.contains("Grumman")==True]
vehicles_df["Make"].replace(["Grumman Allied Industries","Grumman Olson"],"Grumman", inplace=True)

Saleen_names = vehicles_df[vehicles_df["Make"].str.contains("Saleen")==True]
vehicles_df["Make"].replace(["Saleen Performance"],"Saleen", inplace=True)

#vehicles_df["Make"].[vehicles_df.loc("Make"),str.contains("Dutton")]="Dutton"


Chevrolet                             3643
Ford                                  2946
Dodge                                 2360
GMC                                   2347
Toyota                                1836
BMW                                   1677
Mercedes-Benz                         1284
Nissan                                1253
Volkswagen                            1047
Mitsubishi                             950
Mazda                                  915
Audi                                   890
Porsche                                862
Honda                                  836
Jeep                                   829
Pontiac                                784
Subaru                                 781
Volvo                                  717
Hyundai                                662
Chrysler                               641
Buick                                  537
Mercury                                532
Suzuki                                 512
Cadillac   

In [13]:
dutton_filter=vehicles_df["Make"].str.contains("Dutton")

vehicles_df["Make"][dutton_filter]="Dutton"


/var/folders/m3/8y4gy36n5d56ph0nly0_0n180000gp/T/ipykernel_47049/574062409.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  vehicles_df["Make"][dutton_filter]="Dutton"


In [22]:
vehicles_df["Make"].value_counts().index.str.contains("Dutton")


array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False,  True, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False])

In [ ]:
#Convert CO2 Emissions from Grams/Mile to Grams/Km

vehicles_df["CO2 Emission Grams/Mile"]=vehicles_df["CO2 Emission Grams/Mile"]*(1/1.60934)
vehicles_df.rename(columns={"CO2 Emission Grams/Mile":"CO2 Emission Grams/Km"}, inplace=True)
vehicles_df

Converting Grams/Mile to Grams/Km

1 Mile = 1.60934 Km

Grams/Mile * Mile/Km -> Grams/Mile * 1 Mile/1.60934Km

$$ \frac{Grams}{Mile} * \frac{Mile}{Km} $$

$$ \frac{Grams}{Mile} * \frac{1 Mile}{1.60934Km}  $$

In [ ]:
#Create a binary column that solely indicates if the transmission of a car is automatic or manual. Use pandas.Series.str.startswith


vehicles_df["Transmission_Automatic"]=vehicles_df["Transmission"].str.startswith("A", na=False)
vehicles_df

#vehicles_df["Transmission_Automatic"]=list(map(lambda x: "Auto" if(x.startwith("Auto") else "Manual", df["Transmission"])))

In [ ]:
# convert MPG columns to km_per_liter

vehicles_df["City MPG"]=vehicles_df["City MPG"]*(1.60934/3.78541)
vehicles_df["Highway MPG"]=vehicles_df["Highway MPG"]*(1.60934/3.78541)
vehicles_df["Combined MPG"]=vehicles_df["Combined MPG"]*(1.60934/3.78541)
vehicles_df.rename(columns={"City MPG":"City Km/Liter","Highway MPG":"Highway Km/Liter","Combined MPG":"Combined Km/Liter"}, inplace=True)

vehicles_df

convert MPG columns to km_per_liter

MPG = Miles/Gallon -> Km/Liter

1 Mile = 1.60934 Km

1 Gallon = 3.78541 Liters

$$ \frac{Miles}{Gallon} -> \frac{Miles}{Gallon} * \frac{Km}{Miles} * \frac{Gallon}{Liters}$$

$$ \frac{Miles}{Gallon} -> \frac{Miles}{Gallon} * \frac{1.60934Km}{ 1Miles} * \frac{1 Gallon}{3.78541 Liters}$$

* ( 1.60934 / 3.78541 )


### Gathering insights:

- How many car makers are there? How many models? Which car maker has the most cars in the dataset?

- When were these cars made? How big is the engine of these cars?

- What's the frequency of different transmissions, drivetrains and fuel types?

- What's the car that consumes the least/most fuel?

In [ ]:
## How many car makers are there? How many models? Which car maker has the most cars in the dataset?

print(vehicles_df["Make"].nunique())
#127 car makers

print(vehicles_df["Model"].nunique())
#3608 models

print(vehicles_df.Make.value_counts())
#Chevrolet has the most cars

In [ ]:
##When were these cars made? How big is the engine of these cars?

print(vehicles_df["Year"].max(),vehicles_df["Year"].min())
#There were made between 1984 and 2017.

print(vehicles_df["Engine Displacement"].max(),vehicles_df["Engine Displacement"].min())
#Between 0.6  and 8.4 

In [ ]:
##What's the frequency of different transmissions, drivetrains and fuel types?
print(vehicles_df.Transmission_Automatic.value_counts())
print(vehicles_df.Drivetrain.value_counts())
print(vehicles_df["Fuel Type"].value_counts())

In [ ]:
##What's the car that consumes the least/most fuel?

vehicles_df.loc[vehicles_df["Fuel Barrels/Year"] == vehicles_df["Fuel Barrels/Year"].max()]
#Lamborghini Countach cosumes the most 

vehicles_df.loc[vehicles_df["Fuel Barrels/Year"] == vehicles_df["Fuel Barrels/Year"].min()]
#Honda Civic cosumes the least

What brand has the worse CO2 Emissions on average?

Hint: use the function `sort_values()`

In [ ]:
## your Code here

grouped_vehicles_df = vehicles_df.groupby("Make")
mean_df = grouped_vehicles_df.mean()
mean_df.sort_values(by=["CO2 Emission Grams/Km"])

#Vector has the worse CO2 Emissions on average (651.919248)

Do cars with automatic transmission consume more fuel than cars with manual transmission on average?

In [ ]:
## Your Code is here 

grouped_vehicles_df = vehicles_df.groupby("Transmission_Automatic")
mean_df = grouped_vehicles_df.mean()
print(mean_df)

#No, they consume less